Internal compiler error when building Test_DgOperator.cpp with gcc 7.3.0

# Bug reports:

### Expected behavior:

Code compiles

### Current behavior:

When building on ocean with gcc 7.3.0, the compilation fails with this error message. I get this error whether using 1, 6, or 20 threads.

~~~
[ 48%] Building CXX object tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/Test_DgOperator.cpp.o
In file included from /home/geoffrey/Codes/spectre/spectre/tests/Unit/Framework/MockRuntimeSystem.hpp:15:0,
                 from /home/geoffrey/Codes/spectre/spectre/tests/Unit/Framework/ActionTesting.hpp:17,
                 from /home/geoffrey/Codes/spectre/spectre/tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:26:
/home/geoffrey/Codes/spectre/spectre/tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp: In lambda function:
/home/geoffrey/Codes/spectre/spectre/tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:306:43: internal compiler error: in is_normal_capture_proxy, at cp/lambda.c:287
                                           expected_primal_fluxes_vars,
                                           ^
/home/geoffrey/Codes/spectre/spectre/tests/Unit/Framework/TestHelpers.hpp:466:58: note: in definition of macro ‘CHECK_VARIABLES_CUSTOM_APPROX’
         std::decay_t<decltype(a)>, std::decay_t<decltype(b)>>>::apply(a, b,  \
                                                          ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
make[2]: *** [tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/Test_DgOperator.cpp.o] Error 1
make[1]: *** [tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/all] Error 2
make: *** [all] Error 2
~~~

### Environment:
I built this on ocean using the `ocean_gcc.sh` environment.

~~~
CMake Version: 3.18.5
SpECTRE Version: 2021.05.03
Git description: v2021.05.03-61-gcd35535df
Git branch: TimeDependentExportCoords
Git hash: cd35535df1445c0c849afcf56fc2082228fc8ed5
Hostname: ocean.fullerton.edu
Host system: Linux-3.10.0-957.1.3.el7.x86_64
Host system version: 3.10.0-957.1.3.el7.x86_64
Host system processor: x86_64
CMAKE_CXX_FLAGS: 
CMAKE_CXX_LINK_FLAGS:  -L/opt/ohpc/pub/apps/charm_6.10.2_libs/lib -fuse-ld=gold
CMAKE_CXX_FLAGS_DEBUG: -g -DSPECTRE_DEBUG
CMAKE_CXX_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_C_FLAGS: 
CMAKE_C_FLAGS_DEBUG: -g
CMAKE_C_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_Fortran_FLAGS: 
CMAKE_Fortran_FLAGS_DEBUG: -g
CMAKE_Fortran_FLAGS_RELEASE: -O3 -DNDEBUG -O3
CMAKE_EXE_LINKER_FLAGS: 
CMAKE_BUILD_TYPE: Debug
CMAKE_CXX_COMPILER: /opt/ohpc/pub/apps/llvm/10.0.1/bin/g++
CMAKE_CXX_COMPILER_VERSION: 7.3.0
CMAKE_C_COMPILER: /opt/ohpc/pub/apps/llvm/10.0.1/bin/gcc
CMAKE_C_COMPILER_VERSION: 7.3.0
CMAKE_Fortran_COMPILER: /opt/ohpc/pub/compiler/gcc/7.3.0/bin/gfortran
CMAKE_Fortran_COMPILER_VERSION: 7.3.0
Python Version: 3.7.4
~~~

# Feature request:

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Desired feature:

- [ ] Detail 1
- [ ] Detail 2
- [ ] Detail 3

### Detailed discussion:

I found another bug when building with the command `ninja` on Wheeler. 

### Expected behavior:

code compiles 

### Current behavior:

I get this `internal compiler error` trying to build with `ninja`.

```[1/794] Building CXX object tests/Unit/Elliptic/Systems/Poisson/CMakeFiles/Test_Poisson.dir/Test_Tags.cpp.o
[2/794] Building CXX object tests/Unit/Elliptic/Systems/Xcts/CMakeFiles/Test_Xcts.dir/Test_Tags.cpp.o
[3/794] Building CXX object tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/Test_DgOperator.cpp.o
FAILED: tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/Test_DgOperator.cpp.o 
/usr/local/gcc/7.3.0/bin/g++ -DBLAZE_BOOST_THREADS_PARALLEL_MODE=0 -DBLAZE_CPP_THREADS_PARALLEL_MODE=0 -DBLAZE_HPX_PARALLEL_MODE=0 -DBLAZE_MPI_PARALLEL_MODE=0 -DBLAZE_OPENMP_PARALLEL_MODE=0 -DBLAZE_USE_DEFAULT_INITIALIZATON=0 -DBLAZE_USE_OPTIMIZED_KERNELS=1 -DBLAZE_USE_PADDING=0 -DBLAZE_USE_SHARED_MEMORY_PARALLELIZATION=0 -DBLAZE_USE_STREAMING=1 -DBOOST_MULTI_ARRAY_TYPES_RG071801_HPP -DBOOST_SP_DISABLE_THREADS -DDISABLE_OPENBLAS_MULTITHREADING -I../external -I../src -I../tests/Unit -isystem /usr/local/charm/6.10.2/mpi-linux-x86_64-smp-gcc/include -isystem /usr/local/catch/2.13.3 -isystem . -isystem src -isystem src/Parallel -isystem /usr/local/blaze/3.8/include -isystem /usr/local/brigand/master/include -isystem /usr/local/boost/1.65.0-gcc-6.4.0/include -isystem /usr/local/libxsmm/1.16.1/include -isystem /usr/local/yaml-cpp/master/include -isystem /usr/local/python/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/core/include -isystem /usr/local/python/anaconda2-4.1.1/include/python2.7 -isystem /usr/local/libsharp/1.0.0/include -O3 -DNDEBUG -include /home/marlo/spectre/build/SpectrePch.hpp -fdiagnostics-color=always -fPIC -g -march=native -mno-avx512f -ftemplate-backtrace-limit=0 -W -Wall -Wcast-align -Wcast-qual -Wdisabled-optimization -Wextra -Wformat-nonliteral -Wformat-security -Wformat-y2k -Wformat=2 -Winvalid-pch -Wmissing-field-initializers -Wmissing-format-attribute -Wmissing-include-dirs -Wmissing-noreturn -Wno-type-limits -Wnon-virtual-dtor -Wold-style-cast -Woverloaded-virtual -Wpacked -Wpedantic -Wpointer-arith -Wredundant-decls -Wshadow -Wsign-conversion -Wstack-protector -Wswitch-default -Wunreachable-code -Wwrite-strings -Wno-noexcept-type -std=c++1z -D SPECTRE_TEST_REGISTER_FUNCTION=Test_EllipticDG_Test_DgOperator -MD -MT tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/Test_DgOperator.cpp.o -MF tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/Test_DgOperator.cpp.o.d -o tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeFiles/Test_EllipticDG.dir/Test_DgOperator.cpp.o -c ../tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
In file included from ../tests/Unit/Framework/MockRuntimeSystem.hpp:15:0,
                 from ../tests/Unit/Framework/ActionTesting.hpp:17,
                 from ../tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:26:
../tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp: In lambda function:
../tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:306:43: internal compiler error: in is_normal_capture_proxy, at cp/lambda.c:287
                                           expected_primal_fluxes_vars,
                                           ^
../tests/Unit/Framework/TestHelpers.hpp:466:58: note: in definition of macro ‘CHECK_VARIABLES_CUSTOM_APPROX’
         std::decay_t<decltype(a)>, std::decay_t<decltype(b)>>>::apply(a, b,  \
                                                          ^
0x6d9ce0 is_normal_capture_proxy(tree_node*)
    ../../gcc/cp/lambda.c:287
0x6da53c is_lambda_ignored_entity(tree_node*)
    ../../gcc/cp/lambda.c:1254
0x6bfd2f qualify_lookup
    ../../gcc/cp/name-lookup.c:4401
0x6ca0b8 lookup_name_real_1
    ../../gcc/cp/name-lookup.c:5109
0x6ca0b8 lookup_name_real(tree_node*, int, int, bool, int, int)
    ../../gcc/cp/name-lookup.c:5197
0x627f5e cp_parser_lookup_name
    ../../gcc/cp/parser.c:25695
0x644b0b cp_parser_lookup_name_simple
    ../../gcc/cp/parser.c:25754
0x644b0b cp_parser_decltype_expr
    ../../gcc/cp/parser.c:13688
0x644b0b cp_parser_decltype
    ../../gcc/cp/parser.c:13837
0x647fa7 cp_parser_simple_type_specifier
    ../../gcc/cp/parser.c:16742
0x62fa1d cp_parser_type_specifier
    ../../gcc/cp/parser.c:16527
0x6467c7 cp_parser_type_specifier_seq
    ../../gcc/cp/parser.c:20804
0x63d818 cp_parser_type_id_1
    ../../gcc/cp/parser.c:20650
0x63d9d1 cp_parser_template_type_arg
    ../../gcc/cp/parser.c:20730
0x638e61 cp_parser_template_argument
    ../../gcc/cp/parser.c:15935
0x638e61 cp_parser_template_argument_list
    ../../gcc/cp/parser.c:15846
0x638e61 cp_parser_enclosed_template_argument_list
    ../../gcc/cp/parser.c:27031
0x639d4a cp_parser_template_id
    ../../gcc/cp/parser.c:15498
0x63a1f9 cp_parser_class_name
    ../../gcc/cp/parser.c:21976
0x644f95 cp_parser_qualifying_entity
    ../../gcc/cp/parser.c:6291
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
[4/794] Building CXX object tests/Unit/Elliptic/Systems/Xcts/BoundaryConditions/CMakeFiles/Test_XctsBoundaryConditions.dir/Test_Flatness.cpp.o
[5/794] Building CXX object tests/Unit/Elliptic/Utilities/CMakeFiles/Test_EllipticUtilities.dir/Test_ApplyAt.cpp.o
[6/794] Building CXX object tests/Unit/Elliptic/Triggers/CMakeFiles/Test_EllipticTriggers.dir/Test_EveryNIterations.cpp.o
[7/794] Building CXX object tests/Unit/Elliptic/Systems/Poisson/BoundaryConditions/CMakeFiles/Test_PoissonBoundaryConditions.dir/Test_Robin.cpp.o
[8/794] Building CXX object tests/Unit/Elliptic/Systems/Xcts/CMakeFiles/Test_Xcts.dir/Test_Equations.cpp.o
ninja: build stopped: subcommand failed.
```

### Environment:

Building on Wheeler.

```
SpECTRE release version: 2021.05.03
-- The CXX compiler identification is GNU 7.3.0
-- The C compiler identification is GNU 7.3.0
-- The Fortran compiler identification is GNU 7.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/gcc/7.3.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/local/gcc/7.3.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/local/gcc/7.3.0/bin/gfortran - skipped
-- Checking whether /usr/local/gcc/7.3.0/bin/gfortran supports Fortran 90
-- Checking whether /usr/local/gcc/7.3.0/bin/gfortran supports Fortran 90 - yes
-- Found Git: /usr/local/git/2.8.4/bin/git (found version "2.8.4") 
-- Git description: v2021.05.03-82-gfc22b6c
-- Could NOT find Yapf (missing: YAPF_EXECUTABLE YAPF_VERSION) 
-- yapf not available. Please install YAPF (https://github.com/google/yapf)
-- Found Python: /usr/local/python/anaconda2-4.1.1/bin/python2.7 (found version "2.7.12") found components: Interpreter 
-- CMAKE_BUILD_TYPE: Release
-- Found Boost: /usr/local/boost/1.65.0-gcc-6.4.0/include (found suitable version "1.65.1", minimum required is "1.60.0") found components: program_options 
-- Boost include: /usr/local/boost/1.65.0-gcc-6.4.0/include
-- Boost libraries: /usr/local/boost/1.65.0-gcc-6.4.0/lib/libboost_program_options.so
-- Could not find ccache
-- Found Charm: /usr/local/charm/6.10.2/mpi-linux-x86_64-smp-gcc/bin/charmc (found suitable exact version "6.10.2") 
-- Charm++ NOT built with shared memory parallelism
-- Checking for broken std::array<..., 0>
-- Checking for broken std::array<..., 0> -- works
-- Looking for Fortran sgemm
-- Looking for Fortran sgemm - not found
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Looking for Fortran sgemm
-- Looking for Fortran sgemm - found
-- Found BLAS: /usr/local/openblas/0.2.18/lib/libopenblas.so  
-- BLAS libs: /usr/local/openblas/0.2.18/lib/libopenblas.so
-- Disabled OpenBLAS multithreading
-- Looking for Fortran cheev
-- Looking for Fortran cheev - found
-- A library with LAPACK API found.
-- LAPACK libs: /usr/local/openblas/0.2.18/lib/libopenblas.so-lpthread-lm-ldl
-- Found LIBXSMM: /usr/local/libxsmm/1.16.1/include (found suitable version "1.16.1", minimum required is "1.16.1") 
-- LIBXSMM libs: /usr/local/libxsmm/1.16.1/lib/libxsmm.a
-- LIBXSMM incl: /usr/local/libxsmm/1.16.1/include
-- LIBXSMM vers: 1.16.1
-- Found Blaze: /usr/local/blaze/3.8/include (found suitable exact version "3.8") 
-- Blaze incl: /usr/local/blaze/3.8/include
-- Blaze vers: 3.8
-- Found Brigand: /usr/local/brigand/master/include  
-- Brigand include: /usr/local/brigand/master/include
-- Found Catch: /usr/local/catch/2.13.3 (found suitable version "2.13.3", minimum required is "2.8.0") 
-- Catch include: /usr/local/catch/2.13.3
-- Catch version: 2.13.3
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.27.1") 
-- Found GSL: /usr/include (found version "1.15") 
-- GSL libs: /usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so
-- GSL incl: /usr/include
-- GSL vers: 1.15
-- Found HDF5: /usr/local/python/anaconda2-4.1.1/lib/libhdf5.so;/usr/lib64/librt.so;/usr/local/python/anaconda2-4.1.1/lib/libz.so;/usr/lib64/libdl.so;/usr/lib64/libm.so (found version "1.8.16") found components: C 
-- HDF5 libs: /usr/local/python/anaconda2-4.1.1/lib/libhdf5.so/usr/lib64/librt.so/usr/local/python/anaconda2-4.1.1/lib/libz.so/usr/lib64/libdl.so/usr/lib64/libm.so
-- HDF5 incl: /usr/local/python/anaconda2-4.1.1/include
-- HDF5 vers: 1.8.16
-- Using system default memory allocator.
-- Found Python: /usr/local/python/anaconda2-4.1.1/bin/python2.7 (found version "2.7.12") found components: Interpreter Development Development.Module Development.Embed 
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Found pybind11: /usr/local/pybind11/2.6.1/include (found version "2.6.1" )
-- Pybind11 include: /usr/local/pybind11/2.6.1/include
-- Found Libsharp: /usr/local/libsharp/1.0.0/include  
-- libsharp libs: /usr/local/libsharp/1.0.0/lib/libsharp.a
-- libsharp libs: /usr/local/libsharp/1.0.0/lib/libfftpack.a
-- libsharp libs: /usr/local/libsharp/1.0.0/lib/libc_utils.a
-- libsharp incl: /usr/local/libsharp/1.0.0/include
-- Found YAMLCPP: /usr/local/yaml-cpp/master/include  
-- yaml-cpp libs: /usr/local/yaml-cpp/master/lib/libyaml-cpp.so
-- yaml-cpp incl: /usr/local/yaml-cpp/master/include
-- clang-tidy: Not using clang or couldn't find clang-tidy.
-- iwyu: Cannot use include-what-you-use with precompiled header. Pass USE_PCH=OFF to CMake to disable the precompiled header.
-- Found Doxygen: /usr/local/doxygen/1.8.13/bin/doxygen (found version "1.8.13") found components: doxygen dot 
-- Found Python: /usr/local/python/anaconda2-4.1.1/bin/python2.7 (found version "2.7.12") found components: Interpreter 
-- Found PY_bs4: /usr/local/python/anaconda2-4.1.1/lib/python2.7/site-packages/bs4  
-- Could NOT find PY_pybtex (missing: PY_PYBTEX) 
CMake Warning at cmake/SetupDoxygen.cmake:64 (message):
  Doxygen documentation postprocessing is disabled because Python
  dependencies were not found:
Call Stack (most recent call first):
  CMakeLists.txt:127 (include)


CMake Warning at cmake/SetupDoxygen.cmake:71 (message):
  Pybtex missing.  Install with: pip install pybtex
Call Stack (most recent call first):
  CMakeLists.txt:127 (include)


-- Could NOT find PY_coverxygen (missing: PY_COVERXYGEN) 
-- Found Python: /usr/local/python/anaconda2-4.1.1/bin/python2.7 (found version "2.7.12") found components: Interpreter Development Development.Module Development.Embed 
-- Found Python: /usr/local/python/anaconda2-4.1.1/bin/python2.7 (found version "2.7.12") found components: Interpreter 
-- Found NumPy: /usr/local/python/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/core/include (found suitable version "1.11.1", minimum required is "1.10") 
-- NumPy ver. 1.11.1 found (include: /usr/local/python/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/core/include)
-- NumPy incl: /usr/local/python/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/core/include
-- NumPy vers: 1.11.1
-- Found PY_scipy: /usr/local/python/anaconda2-4.1.1/lib/python2.7/site-packages/scipy  
-- 
Useful Information:
-- Git Branch: develop
-- Git Description: v2021.05.03-82-gfc22b6c
-- Git Hash: fc22b6cc9c64d7ebc7f6ffef3252056358673558
-- Build Directory: /home/marlo/spectre/build
-- Source Directory: /home/marlo/spectre
-- Bin Directory: /home/marlo/spectre/build/bin/
-- CMake Modules Path: /home/marlo/spectre/cmake
-- CMAKE_CXX_FLAGS: 
-- CMAKE_CXX_LINK_FLAGS:  -L/usr/local/charm/6.10.2/mpi-linux-x86_64-smp-gcc/lib -fuse-ld=gold
-- CMAKE_CXX_FLAGS_DEBUG: -g -DSPECTRE_DEBUG
-- CMAKE_CXX_FLAGS_RELEASE: -O3 -DNDEBUG
-- CMAKE_EXE_LINKER_FLAGS: 
-- CMAKE_BUILD_TYPE: Release
-- CMAKE_CXX_COMPILER: /usr/local/gcc/7.3.0/bin/g++
-- BUILD_SHARED_LIBS: 
-- USE_PCH: ON
-- ASAN: OFF
-- UBSAN_UNDEFINED: OFF
-- UBSAN_INTEGER: OFF
-- USE_SYSTEM_INCLUDE: ON
-- Python: /usr/local/python/anaconda2-4.1.1/bin/python2.7
-- Python Version: 2.7.12
-- BUILD_PYTHON_BINDINGS: on
-- Doxygen: /usr/local/doxygen/1.8.13/bin/doxygen
```

--- a/tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
+++ b/tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
@@ -302,8 +302,12 @@ void test_dg_operator(
             CAPTURE(vars);
             const auto& primal_fluxes_vars =
                 get_tag(primal_fluxes_vars_tag{}, element_id);
+            // Working around a bug in GCC 7.3 that fails to handle the
+            // structured bindings here
+            const auto& local_expected_primal_fluxes_vars =
+                expected_primal_fluxes_vars;
             CHECK_VARIABLES_CUSTOM_APPROX(primal_fluxes_vars,
-                                          expected_primal_fluxes_vars,
+                                          local_expected_primal_fluxes_vars,
                                           custom_aux_approx);
           }
         }
@@ -319,9 +323,14 @@ void test_dg_operator(
             CAPTURE(vars);
             const auto& operator_applied_to_vars =
                 get_tag(operator_applied_to_vars_tag{}, element_id);
-            CHECK_VARIABLES_CUSTOM_APPROX(operator_applied_to_vars,
-                                          expected_operator_applied_to_vars,
-                                          custom_operator_approx);
+            // Working around a bug in GCC 7.3 that fails to handle the
+            // structured bindings here
+            const auto& local_expected_operator_applied_to_vars =
+                expected_operator_applied_to_vars;
+            CHECK_VARIABLES_CUSTOM_APPROX(
+                operator_applied_to_vars,
+                local_expected_operator_applied_to_vars,
+                custom_operator_approx);
           }
         }
       };

